src/plonkish/backend/halo2.rs

```diff
@@ -357,160 +388,238 @@ pub fn to_halo2_advice<F: PrimeField>(
+                    for (column, values) in assignments.iter() {
+                        let circuit_column = circuit.advice_columns.get(&column.uuid()).unwrap();
+                        let halo2_column = Column::<Any>::from(*circuit_column);
+                        for (offset, value) in values.iter().enumerate() {
+                            assigned_witness[halo2_column.index].as_mut().unwrap()[offset] = *value;
+                        }
+                    }
+                }
+            }
 
-    type Params = Vec<ChiquitoHalo2<F>>;
+            // TODO ignoring the challenges produced by the phase, but can they be useful later?
+            let _ = prover.commit_phase(phase as u8, assigned_witness).unwrap();
+        }
+        prover.create_proof().unwrap();
+        let proof = transcript.finalize();
 
-    fn without_witnesses(&self) -> Self {
-        Self::default()
+        (proof, instance)
     }
+}
 
-    fn params(&self) -> Self::Params {
-        self.sub_circuits.clone()
-    }
+pub trait PlonkishHalo2<F: PrimeField> {
+    fn halo2_setup(&mut self, k: u32, rng: BlockRng<DummyRng>) -> Halo2Setup;
+}
 
-    fn configure_with_params(
-        meta: &mut ConstraintSystem<F>,
-        mut sub_circuits: Self::Params,
-    ) -> Self::Config {
-        sub_circuits
-            .iter_mut()
-            .for_each(|c| c.configure_columns_sub_circuit(meta));
+impl<TG: TraceGenerator<Fr> + Default> PlonkishHalo2<Fr> for PlonkishCompilationResult<Fr, TG> {
+    fn halo2_setup(&mut self, k: u32, rng: BlockRng<DummyRng>) -> Halo2Setup {
+        let mut chiquito_halo2 = ChiquitoHalo2::new(self.circuit.clone());
+        let compiled = compile_halo2_middleware(k, &mut chiquito_halo2).unwrap();
+        // Setup
+        let params = ParamsKZG::<Bn256>::setup::<BlockRng<DummyRng>>(k, rng);
+        let vk = keygen_vk(&params, &compiled).expect("keygen_vk should not fail");
+        let pk = keygen_pk(&params, vk.clone(), &compiled).expect("keygen_pk should not fail");
 
-        let advice_columns: HashMap<UUID, Column<Advice>> =
-            sub_circuits.iter().fold(HashMap::default(), |mut acc, s| {
-                acc.extend(s.advice_columns.clone());
-                acc
-            });
-        let fixed_columns: HashMap<UUID, Column<Fixed>> =
-            sub_circuits.iter().fold(HashMap::default(), |mut acc, s| {
-                acc.extend(s.fixed_columns.clone());
-                acc
-            });
+        Halo2Setup::new(compiled.cs, params, vk, pk, vec![chiquito_halo2])
+    }
+}
 
-        sub_circuits.iter_mut().for_each(|sub_circuit| {
-            sub_circuit.advice_columns = advice_columns.clone();
-            sub_circuit.fixed_columns = fixed_columns.clone();
-            sub_circuit.configure_sub_circuit(meta)
-        });
+impl PlonkishHalo2<Fr> for ChiquitoHalo2SuperCircuit<Fr> {
+    fn halo2_setup(&mut self, k: u32, rng: BlockRng<DummyRng>) -> Halo2Setup {
```

Comment:
@alxkzmn shouldn't this be a method of the plonkish::ir::SuperCircuit? because how do you get to  ChiquitoHalo2SuperCircuit otherwise?

Also, I don't know how can be only one Halo2Setup, as the witness for one is an assigment and for other is a mapping of assignments?

Reply:
I understand your concern about the extra HashMap creation for a single circuit and I will try to fix it. There will be two separate implementations as you suggested. However, Halo2Setup does not include witness, so I don't see a problem here. I believe that it is nice to encapsulate a few things that Halo2 needs regardless of where they are coming. Specifically, CosntraintSystem, Params, PK and VK can live together in a struct.

Reply:
Zooming out, I realized that a better approach could be to only expose a new trait for `SuperCircuit<F, MappingArgs>` - something like `SuperCircuitHalo2Prover` that has `generate_proof()`, and a similar trait `Halo2Prover` for the `PlonkishCompilationResult`. This way, the witness will stay inside the `plonkish::backend::halo2` all the time and the proof generation and verification would look like the following (by the example of or mimc7 circuit):
```rust
    let x_in_value = Fr::from_str_vartime("1").expect("expected a number");
    let k_value = Fr::from_str_vartime("2").expect("expected a number");

    let super_circuit = mimc7_super_circuit::<Fr>();
    let rng = BlockRng::new(DummyRng {});
    let (proof, instance, params, vk) = super_circuit.generate_proof(
      10 /*k, maybe unnecessary if possible to take from the num_rows*/,
      rng,
      (x_in_value, k_value) /*mapping_args*/, 
    )
    let result = halo2_verify(proof, params, vk, instance);
```
Additionally, `halo2_verify(proof, params, vk, instance)` can all be wrapped into a `Halo2Verifier` struct:
```rust
    let x_in_value = Fr::from_str_vartime("1").expect("expected a number");
    let k_value = Fr::from_str_vartime("2").expect("expected a number");

    let super_circuit = mimc7_super_circuit::<Fr>();
    let rng = BlockRng::new(DummyRng {});
    let halo2_verifier = super_circuit.generate_proof(
      10 /*k, maybe unnecessary if possible to take from the num_rows*/,
      rng,
      (x_in_value, k_value) /*mapping_args*/, 
    )
    let result = halo2_verifier.verify();
```
Does this look good to you?

Reply:
We should talk about this. For example, you should not need to generate a proof to verify one.

Reply:
Also, the args cannot be passed to the prover because you might want to generate a evil witness.

Reply:
In this case the witness could be calculated outside, and the verifier could always be constructed from scratch (not by the prover):
```rust
    let x_in_value = Fr::from_str_vartime("1").expect("expected a number");
    let k_value = Fr::from_str_vartime("2").expect("expected a number");

    let super_circuit = mimc7_super_circuit::<Fr>();
    let rng = BlockRng::new(DummyRng {});
    let witness = super_circuit.get_mapping().generate((x_in_value, k_value));
    let (proof, instance, params, vk) = super_circuit.generate_proof(
      10 /*k, maybe unnecessary if possible to take from the num_rows*/,
      rng,
      witness, 
    )
    //-----on the verifier side-------
    let verifier = Halo2Verifier::new(proof, instance, params, vk))
    let result = verifier.verify();
```

Reply:
Alternatively, if evil witness is only necessary for testing, we could expose a method to test-prove with evil witness:
```rust    
    let x_in_value = Fr::from_str_vartime("1").expect("expected a number");
    let k_value = Fr::from_str_vartime("2").expect("expected a number");

    let super_circuit = mimc7_super_circuit::<Fr>();
    let rng = BlockRng::new(DummyRng {});
    let (proof, instance, params, vk) = super_circuit.generate_proof(
      10 /*k, maybe unnecessary if possible to take from the num_rows*/,
      rng,
    ).unwrap();

    let evil_witness = create_evil_witness();
    let failed_proof = super_circuit.test_evil_witness(
      10 /*k, maybe unnecessary if possible to take from the num_rows*/,
      rng,
      evil_witness
    );
    assert!(failed_proof.is_err());

    //-----on the verifier side-------
    let verifier = Halo2Verifier::new(proof, instance, params, vk))
    let result = verifier.verify();

Reply:
@alxkzmn  So let me sumarize my thoughts about the interface:

In the no super circuit scenario:

1. We start with a SBPIR object
2. it should have method to return a plonkish IR thing. I cannot have a method that do halo2 stuff directly because we need to give plonkish config at this step.
3. From the plonkishIR thing we should be able to compile to a halo2 thing, this step cannot be inside another one because we might need to give configuration to the halo2 compilation. This step does not require witness, and the result should have the proving and verifiying keys.
4. The halo2 thing, should have a method to generate a proof, to which we pass plonkish::Assignment. We cannot run the witness generation inside this because we might want to pass an evil witness.

With the Supercircuit in mind, we follow the same process starting with the plonkish::ir::SuperCircuit, and the "things" in the middle can be the same or different type. But the proving method in the halo2 thing should accept a plonkish:SuperAssignments object.

The verification it is correct how it is now, as we shouldn't need to compile anything or generate witness to verify a proof.

I think it is trivial to add num_rows to plonkish IR. Then if you can use that to automatically calculate the halo2 `k` that would be great.

Also the halo2 thingy needs accessors for the proving key, the verifying key, k, etc..

The previous halo2 backend is legacy code now, so feel free to refactor it, or throw it all away.

I think if there are more things, to avoid more cycles perhaps we should jump in a call.